File: Proyectos Intermedios/JavaScript/GestorDeContactosJS/src/services/ContactManager.js
const fs = require('fs').promises;
const path = require('path');
const Contact = require('../models/Contact');

class ContactManager {
  constructor(filePath = path.join(__dirname, '../../data/contacts.csv')) {
    this.filePath = filePath;
    this.contacts = [];
    this.headers = ['id', 'name', 'phone', 'email', 'address', 'createdAt', 'updatedAt'];
  }

  async init() {
    try {
      await fs.access(this.filePath);
    } catch {
      await fs.mkdir(path.dirname(this.filePath), { recursive: true });
      await fs.writeFile(this.filePath, `"${this.headers.join('","')}"\n`, 'utf8');
      console.log('✅ Archivo CSV creado en:', this.filePath);
    }
    await this.loadContacts();
  }

  async loadContacts() {
    try {
      const data = await fs.readFile(this.filePath, 'utf8');
      const lines = data.split('\n').filter(line => line.trim() && !line.startsWith('"id"'));
      
      this.contacts = lines.map(line => Contact.fromCSV(line));
      console.log(`📥 ${this.contacts.length} contactos cargados`);
    } catch (error) {
      console.error('❌ Error cargando contactos:', error.message);
      this.contacts = [];
    }
  }

  async saveContacts() {
    try {
      const csvContent = [
        `"${this.headers.join('","')}"`,
        ...this.contacts.map(contact => contact.toCSV())
      ].join('\n') + '\n';

      await fs.writeFile(this.filePath, csvContent, 'utf8');
      console.log(`💾 ${this.contacts.length} contactos guardados`);
    } catch (error) {
      console.error('❌ Error guardando contactos:', error.message);
      throw error;
    }
  }

  async addContact(name, phone, email, address = '') {
    const contact = new Contact(null, name, phone, email, address);
    const validation = contact.isValid();
    
    if (!validation.valid) {
      throw new Error('Validación fallida:\n- ' + validation.errors.join('\n- '));
    }

    const exists = this.contacts.some(c => 
      c.email === contact.email || c.phone === contact.phone
    );

    if (exists) {
      throw new Error('⚠️ Ya existe un contacto con este email o teléfono');
    }

    this.contacts.push(contact);
    await this.saveContacts();
    return contact;
  }

  searchContacts(query) {
    const q = query.toLowerCase().trim();
    return this.contacts.filter(contact => 
      contact.name.toLowerCase().includes(q) ||
      contact.phone.includes(q) ||
      contact.email.includes(q) ||
      contact.address.toLowerCase().includes(q)
    );
  }

  getAllContacts() {
    return [...this.contacts].sort((a, b) => a.name.localeCompare(b.name));
  }

  async updateContact(id, updates) {
    const contact = this.contacts.find(c => c.id === id);
    if (!contact) throw new Error('❌ Contacto no encontrado');

    contact.update(updates);
    const validation = contact.isValid();
    if (!validation.valid) {
      throw new Error('Validación fallida:\n- ' + validation.errors.join('\n- '));
    }

    await this.saveContacts();
    return contact;
  }

  async deleteContact(id) {
    const index = this.contacts.findIndex(c => c.id === id);
    if (index === -1) throw new Error('❌ Contacto no encontrado');

    const deleted = this.contacts.splice(index, 1)[0];
    await this.saveContacts();
    return deleted;
  }

  getContactById(id) {
    return this.contacts.find(c => c.id === id);
  }
}

module.exports = ContactManager;

File: Proyectos Intermedios/Java/GestorDeContactosJava/src/main/java/service/ContactManager.java
package service;

import model.Contact;

import java.io.*;
import java.nio.file.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Collectors;

public class ContactManager {
    private static final String[] HEADERS = {"id", "name", "phone", "email", "address", "created_at", "updated_at"};
    private final Path csvPath;
    private final List<Contact> contacts = new ArrayList<>();

    public ContactManager(String csvPath) {
        this.csvPath = Paths.get(csvPath);
        ensureDataDirectory();
        loadContacts();
    }

    private void ensureDataDirectory() {
        try {
            Files.createDirectories(csvPath.getParent());

            if (!Files.exists(csvPath)) {
                try (BufferedWriter writer = Files.newBufferedWriter(csvPath, StandardCharsets.UTF_8)) {
                    writer.write(String.join(",", HEADERS));
                    writer.newLine();
                }
                System.out.println("✅ Archivo CSV creado en: " + csvPath.toAbsolutePath());
            }
        } catch (IOException e) {
            System.err.println("❌ Error creando directorio de datos: " + e.getMessage());
        }
    }

    public void loadContacts() {
        contacts.clear();
        try (BufferedReader reader = Files.newBufferedReader(csvPath, StandardCharsets.UTF_8)) {
            String line;
            boolean isFirstLine = true;

            while ((line = reader.readLine()) != null) {
                if (isFirstLine) {
                    isFirstLine = false;
                    continue; // Saltar header
                }
                if (line.trim().isEmpty()) continue;

                try {
                    Contact contact = Contact.fromCsvRow(line);
                    contacts.add(contact);
                } catch (Exception e) {
                    System.err.println("⚠️ Error cargando contacto: " + e.getMessage());
                }
            }

            System.out.println("📥 " + contacts.size() + " contactos cargados desde: " + csvPath.getFileName());
        } catch (IOException e) {
            System.err.println("ℹ️ Archivo CSV no encontrado, se creará al guardar el primer contacto");
        }
    }

    public void saveContacts() {
        try (BufferedWriter writer = Files.newBufferedWriter(csvPath, StandardCharsets.UTF_8)) {
            writer.write(String.join(",", HEADERS));
            writer.newLine();

            for (Contact contact : contacts) {
                writer.write(contact.toCsvRow());
                writer.newLine();
            }

            System.out.println("💾 " + contacts.size() + " contactos guardados");
        } catch (IOException e) {
            System.err.println("❌ Error guardando contactos: " + e.getMessage());
        }
    }

    public Contact addContact(String name, String phone, String email, String address) throws Exception {
        // Generar ID único
        String contactId = java.util.UUID.randomUUID().toString().substring(0, 10);
        Contact contact = new Contact(contactId, name, phone, email, address);

        // Validar
        Contact.ValidationResult result = contact.isValid();
        if (!result.isValid()) {
            throw new Exception(result.getErrors());
        }

        // Verificar duplicados
        boolean exists = contacts.stream().anyMatch(c ->
                c.getEmail().equalsIgnoreCase(contact.getEmail()) ||
                        c.getPhone().equals(contact.getPhone())
        );

        if (exists) {
            throw new Exception("⚠️ Ya existe un contacto con este email o teléfono");
        }

        contacts.add(contact);
        saveContacts();
        return contact;
    }

    public List<Contact> searchContacts(String query) {
        if (query == null || query.trim().isEmpty()) {
            return new ArrayList<>(contacts);
        }

        String q = query.toLowerCase().trim();
        return contacts.stream()
                .filter(c -> c.getName().toLowerCase().contains(q) ||
                        c.getPhone().contains(q) ||
                        c.getEmail().contains(q) ||
                        c.getAddress().toLowerCase().contains(q))
                .collect(Collectors.toList());
    }

    public List<Contact> getAllContacts() {
        return contacts.stream()
                .sorted(Comparator.comparing(c -> c.getName().toLowerCase()))
                .collect(Collectors.toList());
    }

    public Contact getContactById(String id) {
        return contacts.stream()
                .filter(c -> c.getId().equals(id))
                .findFirst()
                .orElse(null);
    }

    public Contact updateContact(String id, String name, String phone, String email, String address) throws Exception {
        Contact contact = getContactById(id);
        if (contact == null) {
            throw new Exception("❌ Contacto no encontrado");
        }

        contact.update(name, phone, email, address);

        // Validar actualización
        Contact.ValidationResult result = contact.isValid();
        if (!result.isValid()) {
            throw new Exception(result.getErrors());
        }

        saveContacts();
        return contact;
    }

    public Contact deleteContact(String id) throws Exception {
        Contact contact = getContactById(id);
        if (contact == null) {
            throw new Exception("❌ Contacto no encontrado");
        }

        contacts.remove(contact);
        saveContacts();
        return contact;
    }
}